### PR TITLE
Asset validation

### PR DESCRIPTION
**Pull request by tokejepsen**
- Asset, version and components are now validated in order.
- Version creation happens in the validator automaticly, so the logic/code has just been moved from the selector to the validator
- Asset is checking against ftrackAssetName in the context

### CODE REVIEW

**Comment by mottosso** [comment]
> Version creation happens in the validator automaticly, so the logic/code has just been moved from the selector to the validator

This sounds dangerous, Validators should be idempotent and should not modify or create anything. 

What about having it happen during Extraction or Conform?

**Comment by mkolar** [comment]
This is a tricky one, because to be able to successfully publish to ftrack we do need to check all three levels of an AssetVersion and make sure they exist before moving onto checking the next thing.

Until this pull request it was scattered across all stages. Asset checked and created in selector, version and component checked in validation and created in conform. Having it all in validation at least gives the safety net of stopping the publish if something goes wrong. Also it gives us the option to use fully filled out ftrackData in extractors and conformers of other extensions (deadline for example).

So for me, unless we run into some serious problems, it seems like the least painful option.

**Comment by mottosso** [comment]
> This is a tricky one, because to be able to successfully publish to ftrack we do need to check all three levels of an AssetVersion and make sure they exist before moving onto checking the next thing.

Ok, I think I have an idea of what needs to happen, but you're not gonna like it. ;)

I think what needs to happen is for FTrack to be populated _before_ publishing. To make that part of an artists responsibility, either by hand, or some utility written for him.

That way, publishing can be about validating what's already there, and less about manipulating data. It sounds like the FTrack entry is as relevant to an artists work as the content itself.

And again, I realise nobody is keen on the idea of adding additional workload to artists, especially when they've already gotten to this point at which they don't have to. But I suspect it will add workload to you both further down the line in the form of debugging and Q&A. :S

**Comment by mottosso** [comment]
Or at the very least, warn about it, and give them the option to repair.

The bottom line being, implicitly modifying anything without anyone's knowledge is never a good thing.

**Comment by mkolar** [comment]
Separating version creation from publishing process would cancel out most of the advantages of publishing in the first place. Quite frankly if I had to force artist to deal with it on their own, I might as well not have them publishing at all. We've been publishing by hand until now and it cause s more screw ups then mistakes in work files themselves. 

We could potentially split this into a validator and a few extractors/conformers, however we would lose quite a bit of flexibility in terms of sharing this information with other extensions that might benefit from it. 

For example deadline extension needs ftrack asset to be created before the extraction. So to sort this we'd need to run them in order

**Comment by mottosso** [comment]
> For example deadline extension needs ftrack asset to be created before the extraction. So to sort this we'd need to run them in order

If the ftrack entry is created before any plug-ins are run, wouldn't this still work out though?

> We could potentially split this into a validator and a few extractors/conformers

At the very least, yes. At least it can be expected of an Extractor or Conformer to produce data which is a step in the right direction I think. You would loose the ability to validate it, but I think it's safe to assume that anything done automatically is never wrong.

**Comment by mkolar** [comment]
@tokejepsen how about we try doing it in 1 validator and 1 extractor?

validator checks for asset, if it exists checks for version, if it exists, checks components. We know that if asset doesn't exist, we don't need to be checking the other two hence I think it can all stay in the same validator.

Extractor (which would ideally be always ran before deadline extractors) then check whether it needs, to create asset and version or only version based on a tag we pass it from the validator. that should keep the functionality, stay more within the boundaries of SVEC and keep it possible to run other extensions dependent on it (provided, they are offset from the base extractor order.)

**Comment by tokejepsen** [comment]
> Do you thing we could do with just checking if component with the same name exists?

I guess we could do, though what would happen when the component already exists?
I don't want to fail, since a pre-existing component is the most common case. So then its a question of what happens when their values aren't the same. Do we just overwrite?

> I think we should stick this ordering of plugins a bit closer together, to make it a bit clearer they belong together. How about +0.11, +0.12 and +0.13.

Sure:)

> how about we try doing it in 1 validator and 1 extractor?

Sounds good to me, will have a look at this today hopefully and update this pull-request

**Comment by mottosso** [comment]
> validator checks for asset, if it exists checks for version, if it exists, checks components. We know that if asset doesn't exist, we don't need to be checking the other two hence I think it can all stay in the same validator.
> 
> Extractor (which would ideally be always ran before deadline extractors) then check whether it needs, to create asset and version or only version based on a tag we pass it from the validator. that should keep the functionality, stay more within the boundaries of SVEC and keep it possible to run other extensions dependent on it (provided, they are offset from the base extractor order.)

+1

**Comment by mkolar** [comment]
> So then its a question of what happens when their values aren't the same. Do we just overwrite?

If we're conforming the file to be published to some other location before uploading it as a component, values will never be the same. I reckon we should throw a warning in the validation, that this is happening, and tell user, that this component will not be touched during this publish. They are trying to publish into the same version after all which means overwriting, which I don't like at all. If they need to fix on component out of 3 for example, I reckon they should delete the component first and then run the publish. That was anything already existing will stay untouched and the new component will get appended to it.

To put it simply

``` python
if componentName == onlineComponentName:
    warning('this component already exists. If you want to replace it, please delete it in the webUI first')
    pass
```

**Comment by tokejepsen** [comment]
> That was anything already existing will stay untouched and the new component will get appended to it.

Cool, warning it is.

**Comment by mkolar** [comment]
There are a few small things I had to quickly take care of in my fork. Maybe it would be for the best if you could include them in your pull request, so we're not bouncing the code around too much (I have some ad-hoc changes that I don't want to be cleaning up now.)

[publishing version if invisible in webUI](https://github.com/pyblish/pyblish-ftrack/blob/a1603b6f5a646547c63d95bf9d4f10d48c648317/pyblish_ftrack/plugins/validate_ftrack_version.py#L27-L29) if we just throw an error that the version is invisible, we get into an infinite loop. Artist can't remove the version because he can't see it, and he can't publish a new one, because it already exists. I reckon if we run into invisible version, we publish it, so artist can deal with it.

[select_ftrack_data](https://github.com/pyblish/pyblish-ftrack/blob/a1603b6f5a646547c63d95bf9d4f10d48c648317/pyblish_ftrack/plugins/select_ftrack_data.py#L60) - we should change refer to the parent of the asset as 'parent' or 'entity' to be ready for potential publishing to asset builds, not just shots. I've tried doing asset build publishing and it will only need very minor tweaks.

[reviewable component](https://github.com/pyblish/pyblish-ftrack/blob/a1603b6f5a646547c63d95bf9d4f10d48c648317/pyblish_ftrack/plugins/conform_ftrack_component.py#L40-L41) I've added it in my repo, but again would be nice if you could just include the few liner in your pull request so we can get on the same page.

**Comment by tokejepsen** [comment]
cool, Ill update my repo with your fork

**Comment by tokejepsen** [comment]
closing this in favour of new pull request pending
